### ISSUE TITLE
Editor line misalignment

### ISSUE THREAD

**Issue by Daeraxa** [pulsar-edit/pulsar#355]
### Thanks in advance for your bug report!

- [X] Have you reproduced issue in [safe mode](https://pulsar-edit.dev/docs/atom-archive/faq/sections/what-does-safe-mode-do.html)?
- [X] Have you used the [debugging guide](https://pulsar-edit.dev/docs/atom-archive/hacking-atom/#debugging) to try to resolve the issue?
- [X] Have you checked our [FAQs](https://pulsar-edit.dev/docs/launch-manual/sections/faq/) to make sure your question isn't answered there?
- [X] Have you checked to make sure your issue does not already [exist](https://github.com/issues?q=is%3Aissue+user%3Apulsar-edit+is%3Aopen)?
- [X] Have you checked you are on the latest release of Pulsar?

### What happened?

This is a weird issue I've seen a whole bunch of times, it seems it was an issue in Atom too (well it is because I just replicated this in 1.60.0 too) but it seems it was always hard to nail down and reliably replicate.

I randomly stumbled upon an exact combination that replicate this, you can see the effect of it here, note how the lines don't align properly horizontally and are also shifted by quite a way leading to a "staggered" effect between the editor and line numbers.

![image](https://user-images.githubusercontent.com/58074586/215295278-753a84b0-8bf8-416c-b22f-b497962e3234.png).

Replication steps and code below.

### Pulsar version

1.101.0-dev (but also in Atom 1.60.0)

### Which OS does this happen on?

🪟 Windows

### OS details

10 but also in Fedora 36

### Which CPU architecture are you running this on?

64-bit(x86_64)

### What steps are needed to reproduce this?

- Open a new, empty Pulsar session with an empty, plain text buffer displayed, default core packages only.
- Paste in the code you can find in "Additional information" below
- Scroll to the top of the file
- Change the editor language from `Plain Text` to `HTML`
- Observe issue

### Additional Information:

The code in question (yes it looks weird and that is because it is an internal... thing... for one of my company's products, basically a configurable dynamic page header/report thingy that uses HTML mixed with our internal scripting language and database calls).

```html
<html>
  <head>
       <style type="text/css">
body ~{
       color: black;
       font-size: 11;
       font-family: "Verdana";
       background-color: #fbfbfb;
       border: 1px solid #c0c0c0;
       overflow-x: hidden;
       }
    img ~{
       border: 0;
       }
    a ~{
       color: blue;
       font-size: 13;
       font-family: "Verdana";
       }
    td ~{
       font-size: 11;
       vertical-align: top;
       }
    </style>
    <title></title>
  </head>
  <body>
     <div id="basic">
      <table style="width: 100%;">
        <tr>
          <td style="width: 30%;">
            <strong>Sample Id:</strong> {=.Identifier}
          </td>
          <td style="width: 20%;">
            <strong>Priority:</strong> {=ToString("SamplePriority",.Priority)}
          </td>
          <td style="width: 20%;">
            <strong>Age:</strong> {=IfKnownString(ToString(Truncate(.Patient.AgeInYears(DateTimeToDate(.CreationTime)),0),  "%g"))}
          </td>
          <td style="width: 30%;">
            <strong>Patient Id:</strong> {=IfKnownString(.Patient.Identifier)}
          </td>
        </tr>
        <tr>
          <td>
            <strong>Patient name:</strong> {=
            IfKnownString(.Patient.LastName)}{=IF .Patient.MiddleName <> ? THEN ", " + .Patient.MiddleName ELSE ""}{=IF  .Patient.FirstName
            <> ? THEN ", " + .Patient.FirstName ELSE ""}
          </td>
          <td>
            <strong>Sex:</strong> {= IfKnownString(ToString("Sex",.Patient.Sex))}
          </td>
          <td>
            <strong>DOB:</strong> {= IfKnownString(ToString(.Patient.BirthDate,"%x"))}
          </td>
          <td>
            <strong>Patient location:</strong> {=IfKnownString(.PatientLocation.Name)}
          </td>
        </tr>
        </table>
    </div>
   </body>
</html>
```

**Comment by confused-Techie** [comment]
Super awesome you were able to nail down the exact steps needed to reproduce this. But have a feeling it'll be just as hard to nail down why it happens but this is a step in the right direction

**Comment by Spiker985** [comment]
So I just found out that this error seems to occur because the content contained within the `document.querySelector("body > atom-workspace > atom-workspace-axis > atom-workspace-axis > atom-pane-container > atom-pane > div > atom-text-editor:nth-child(1) > div > div.scroll-view")` element is simply offset on the screen

I ran into the problem natively, and by going into dev tools navigating to `document.querySelector("body > atom-workspace > atom-workspace-axis > atom-workspace-axis > atom-pane-container > atom-pane > div > atom-text-editor:nth-child(1) > div > div.scroll-view > div.vertical-scrollbar")`, right-click, Scroll into View

My content was shown again as expected. Obviously the above selectors are relative to the tab index and other things, but should hopefully provide some assistance in determining _why_ this happens.

My problem came up when I was repeatedly using the `editor:toggle-soft-wrap` command, and scrolling the editing pane

**Comment by bobemoe** [comment]
Yes i've noticed this too. Happens for me occasionally, feels like it happens after a external change to the file content. I think I have it occur after git has made a large change. Currently beginning to notice it more after doing an "Editor: Auto Indent". Just managed to capture one: 

![image](https://user-images.githubusercontent.com/2304970/220765260-7b6709bd-06bc-41c8-8504-cc70c8c36f89.png)

Its not just a slight pixel misalignment like it might look like on the screenshot, its actually line 10 lines off compared to the actual code:

![image](https://user-images.githubusercontent.com/2304970/220765384-4c4abb99-015b-4739-8cd8-be7a3d3be0f4.png)

I do use auto indent a lot. This one happened after doing a whole file, only 66 lines but entire file. i probably did a select all first.

**Comment by savetheclocktower** [comment]
I've seen this happen a few times but can't reproduce using @Daeraxa's instructions.

I wonder if it has anything to do with the error I get on window load maybe ~10% of the time:

```
TypeError: Cannot read property 'element' of undefined
    at TextEditorComponent.didScrollDummyScrollbar (/Applications/Pulsar.app/Contents/Resources/app.asar/src/text-editor-component.js:1769:37)
```

**Comment by Spiker985** [comment]
Ironically enough, I also took a look at this about an hour ago on my work machine and was unable to replicate it on 1.101.0 (there were some non-standard packages, however)

I wonder if there are different conditions per OS which is why this is so elusive

It's also confusing to me that it isn't an error, but instead a miscalculation (maybe)

**Comment by Daeraxa** [comment]
> I wonder if there are different conditions per OS which is why this is so elusive

That is why I tested it on two unrelated machines before logging it. One was my work laptop on Windows and the other was my own on Fedora, both with newly generated .pulsar config folders.

**Comment by Spiker985** [comment]
You know, I do have 3 brand new, fresh installs of Big Sur, Monterey and Ventura that I can try to replicate this on - let's give it a whirl

**Comment by Spiker985** [comment]
This is a broken one (which was actually recreated using Dae's instructions). I pasted the content, didn't move the cursor in the buffer, scrolled up to line 1, and then changed the language via Ctrl + Shift + L
![ventura_scroll_view_lines_translate](https://user-images.githubusercontent.com/7829451/220796828-7e29fa45-9a3c-4262-9f36-0b1f7efebc10.gif)

This is scrolling some standard working text
![arch_linux_scroll_view_lines_translate](https://user-images.githubusercontent.com/7829451/220796838-9ef84dff-36b3-4166-81ac-56315a2994ca.gif)

Something to note about the broken one, the broken div contains other divs with data-screen-row attributes which appear to be "stuck"
![ventura_scroll_view_stuck_data-screen-row](https://user-images.githubusercontent.com/7829451/220798116-4832b091-20fe-420e-b63a-34a8b3f3e260.gif)

I think there's an edge case in either `src/text-editor.js`, `src/text-editor-element.js`, or `src/text-editor-component.js` where it is improperly calculating how much it needs to render.

It's not the scroll event, because it does scroll _properly_, the content of the render is the problem.

The other thing to think about is that the line numbers, and the text buffers are rendered independently - because you can disable line numbers _per text editor_

I haven't yet tested if you toggle `atom.textEditors.getActiveTextEditor().setLineNumberGutterVisible(<bool>)` while it is broken, if the line numbers will sync back up with the broken renderer or not

**Comment by bobemoe** [comment]
OK, so I've worked out how to reproduce my case fairly consistently now using auto indent. Here are my steps:

1. Open a file (i have my project open and file tree pane on the left) it needs to be one that is longer than the screen. 
2. Highlight a block of code, shift-tab a few times to remove the indents.
3. Keeping the block highlighted, scroll so that the block is completely off the screen.
4. Auto indent. (ctrl-shit p -> Editor: Auto Indent)

Seems to break every time for me using this method.

**Comment by bobemoe** [comment]
pretty sure this is fixed by https://github.com/pulsar-edit/pulsar/pull/798

**Comment by savetheclocktower** [comment]
Yep, should be. @Daeraxa can reopen it if somehow it's a different issue.
